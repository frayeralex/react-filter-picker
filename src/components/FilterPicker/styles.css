.FilterPiker {
  width: 300px;
  position: absolute;
  top: 0;
  left: 0;
  background: #f1f1f1;
  transition: .4s;
  animation: fadeIn .3s;
  transform-origin: bottom center;

}

.FilterPiker_triangular {
  position: absolute;
  width: 0;
  height: 0;
  border-left: 10px solid transparent;
  border-right: 10px solid transparent;
  border-bottom: 10px solid rgba(0, 0, 0, .1);
  animation: fadeIn .3s;
  transform-origin: bottom center;
  transition: .2s;
}

.FilterPiker_content {
  display: flex;
  flex-direction: column;
}

@keyframes fadeIn {
  from {
    opacity: 0;
    transform: translateY(20px);
  }
  to {
    transform: translateY(0)
  }
}
