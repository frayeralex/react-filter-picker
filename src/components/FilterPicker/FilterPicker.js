import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MobileDetect from 'mobile-detect';
import { CollapsedBox, FilterGroup } from '../index';
import { filterData, selectedFilters } from '../../types';
import styles from './styles.css';

class FilterPicker extends Component {
  static propTypes = {
    filterData: filterData,
    selectedFilters: selectedFilters,
    activeFilters: PropTypes.arrayOf(PropTypes.string),
    onApply: PropTypes.func,
    coordinates: PropTypes.shape({
      top: PropTypes.number,
      bottom: PropTypes.number,
      left: PropTypes.number,
      right: PropTypes.number,
      width: PropTypes.number,
      height: PropTypes.number,
    }),
    md: PropTypes.shape({
      tablet: PropTypes.func,
      mobile: PropTypes.func,
    }),
  };

  static defaultProps = {
    onApply: () => {},
    activeFilters: [],
    selectedFilters: {},
    md: new MobileDetect(window.navigator.userAgent),
  };

  constructor(props) {
    super(props);

    this.state = {
      selectedFilters: props.selectedFilters,
    };
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.activeFilters !== this.props.activeFilters) {
      this.setState({ selectedFilters: this.props.selectedFilters });
    }
  }

  handleApplyClick = () => {
    this.props.onApply(this.state.selectedFilters);
  };

  getStyles() {
    const { coordinates } = this.props;
    const isDesktop = !this.props.md.phone() && !this.props.md.tablet();
    if (!isDesktop) {
      return {
        width: 'calc(100% - 20px)',
        left: 10,
        top: coordinates.top + coordinates.height + 10,
      };
    }
    const width = 300;
    const left = coordinates.left + coordinates.width / 2 - width / 2;

    return {
      width,
      left: left <= 10 ? 10 : left,
      top: coordinates.top + coordinates.height + 10,
    };
  }

  getTriangularStyles() {
    return {
      top: this.props.coordinates.top + this.props.coordinates.height,
      left: this.props.coordinates.left + this.props.coordinates.width / 2 - 10,
    };
  }

  handleContainerClick = event => {
    event.stopPropagation();
    event.nativeEvent.stopImmediatePropagation();
  };

  handleApplyFilterGroup = (filter, selectedFilterIdList) => {
    this.setState(
      {
        selectedFilters: {
          ...this.state.selectedFilters,
          [filter]: selectedFilterIdList,
        },
      },
      this.handleApplyClick,
    );
  };

  render() {
    return (
      <React.Fragment>
        <span
          style={{ ...this.getTriangularStyles() }}
          className={styles.FilterPiker_triangular}
        />
        <div
          onClick={this.handleContainerClick}
          style={{ ...this.getStyles() }}
          className={styles.FilterPiker}
        >
          <div className={styles.FilterPiker_content}>
            {this.props.activeFilters.length > 0 &&
              this.props.activeFilters.map((filter, index) => {
                return (
                  <CollapsedBox
                    expanded={!index}
                    key={filter}
                    title={this.props.activeFilters.length > 1 ? filter : null}
                  >
                    <FilterGroup
                      filter={filter}
                      filterItems={this.props.filterData[filter]}
                      selectedFilters={this.state.selectedFilters[filter]}
                      onApply={this.handleApplyFilterGroup}
                    />
                  </CollapsedBox>
                );
              })}
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export default FilterPicker;
