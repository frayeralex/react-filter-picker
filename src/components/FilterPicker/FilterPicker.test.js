import React from 'react';
import FilterPicker from './FilterPicker';
import { CollapsedBox, FilterGroup } from '../index';

const phone = jest.fn();
const tablet = jest.fn();

const props = {
  title: 'title',
  md: {
    phone,
    tablet,
  },
  filterData: {
    color: [
      {
        id: '1',
        title: 'first',
      },
      {
        id: '2',
        title: 'second',
      },
    ],
    size: [
      {
        id: '1',
        title: 'first',
      },
      {
        id: '2',
        title: 'second',
      },
    ],
  },
  activeFilters: ['color'],
  coordinates: {
    top: 10,
    bottom: 10,
    left: 10,
    right: 10,
    width: 100,
    height: 60,
  },
};

describe('<FilterPicker/>', () => {
  it('should render without crashing', () => {
    const wrapper = shallow(<FilterPicker {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should correct render with more then 1 active filter', () => {
    const wrapper = shallow(<FilterPicker {...props} />);
    const activeFilters = ['color', 'size'];
    expect(wrapper.find(CollapsedBox)).toHaveLength(1);
    expect(wrapper.find(FilterGroup)).toHaveLength(1);
    wrapper.setProps({ activeFilters });
    expect(wrapper.find(CollapsedBox)).toHaveLength(2);
    expect(wrapper.find(FilterGroup)).toHaveLength(2);
    expect(wrapper).toMatchSnapshot();
  });
  it('should handle goupApply click correctly', () => {
    const onApply = jest.fn();
    const expectSelectedFilters = {
      color: ['1'],
    };
    const wrapper = shallow(<FilterPicker {...props} onApply={onApply} />);
    wrapper
      .find(FilterGroup)
      .first()
      .props()
      .onApply('color', ['1']);
    expect(wrapper.state()).toEqual({
      selectedFilters: expectSelectedFilters,
    });
    expect(onApply).toHaveBeenCalledWith(expectSelectedFilters);
  });
  it('should correct handle root element click', () => {
    const stopPropagation = jest.fn();
    const stopImmediatePropagation = jest.fn();
    const wrapper = shallow(<FilterPicker {...props} />);
    wrapper
      .find('div')
      .first()
      .simulate('click', {
        stopPropagation,
        nativeEvent: {
          stopImmediatePropagation,
        },
      });
    expect(stopPropagation).toHaveBeenCalled();
    expect(stopImmediatePropagation).toHaveBeenCalled();
  });
  it('should show 300px width dropdown on desktops', () => {
    const md = {
      phone: jest.fn(() => false),
      tablet: jest.fn(() => false),
    };
    const wrapper = shallow(<FilterPicker {...props} md={md} />);
    expect(
      wrapper
        .find('div')
        .first()
        .prop('style').width,
    ).toEqual(300);
  });
  it('should show fullwidht dropdown on phones', () => {
    const md = {
      phone: jest.fn(() => true),
      tablet: jest.fn(() => false),
    };
    const wrapper = shallow(<FilterPicker {...props} md={md} />);
    expect(
      wrapper
        .find('div')
        .first()
        .prop('style').width,
    ).toEqual('calc(100% - 20px)');
  })
  it('should show fullwidht dropdown on tablet', () => {
    const md = {
      phone: jest.fn(() => false),
      tablet: jest.fn(() => true),
    };
    const wrapper = shallow(<FilterPicker {...props} md={md} />);
    expect(
      wrapper
        .find('div')
        .first()
        .prop('style').width,
    ).toEqual('calc(100% - 20px)');
  })
});
