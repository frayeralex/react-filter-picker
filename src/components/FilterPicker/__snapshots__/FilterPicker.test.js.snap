// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterPicker/> should correct render with more then 1 active filter 1`] = `
<Fragment>
  <span
    style={
      Object {
        "left": 50,
        "top": 70,
      }
    }
  />
  <div
    onClick={[Function]}
    style={
      Object {
        "left": 10,
        "top": 80,
        "width": 300,
      }
    }
  >
    <div>
      <CollapsedBox
        expanded={true}
        key="color"
        title="color"
      >
        <FilterGroup
          filter="color"
          filterItems={
            Array [
              Object {
                "id": "1",
                "title": "first",
              },
              Object {
                "id": "2",
                "title": "second",
              },
            ]
          }
          onApply={[Function]}
          selectedFilters={Array []}
        />
      </CollapsedBox>
      <CollapsedBox
        expanded={false}
        key="size"
        title="size"
      >
        <FilterGroup
          filter="size"
          filterItems={
            Array [
              Object {
                "id": "1",
                "title": "first",
              },
              Object {
                "id": "2",
                "title": "second",
              },
            ]
          }
          onApply={[Function]}
          selectedFilters={Array []}
        />
      </CollapsedBox>
    </div>
  </div>
</Fragment>
`;

exports[`<FilterPicker/> should render without crashing 1`] = `
<Fragment>
  <span
    style={
      Object {
        "left": 50,
        "top": 70,
      }
    }
  />
  <div
    onClick={[Function]}
    style={
      Object {
        "left": 10,
        "top": 80,
        "width": 300,
      }
    }
  >
    <div>
      <CollapsedBox
        expanded={true}
        key="color"
        title={null}
      >
        <FilterGroup
          filter="color"
          filterItems={
            Array [
              Object {
                "id": "1",
                "title": "first",
              },
              Object {
                "id": "2",
                "title": "second",
              },
            ]
          }
          onApply={[Function]}
          selectedFilters={Array []}
        />
      </CollapsedBox>
    </div>
  </div>
</Fragment>
`;
